vpr/src/place/noc_place_utils.h

```diff
@@ -0,0 +1,375 @@
+ * costs get really low. So we expect the latency costs to never actually go 
+ * below one picosecond, so we check whether a recomputed cost is below this
+ * threshold and if it is then indicate that there is no need to verify the
+ * NoC latency cost. This is acceptable since the only case where the NoC
+ * latency cost will be below one picosecond is when the weighting on the 
+ * latency component of the cost is extremely low (ie. 0) and when most if not 
+ * all the latency constraints have been met (so this is also close to 0). There
+ * is no need to check and compare the cost here.
+ * 
+ */
+bool check_recomputed_noc_latency_cost(float recomputed_cost);
+
+void allocate_and_load_noc_placement_structs(void);
+
+void free_noc_placement_structs(void);
+
+/* Below are functions related to the feature that forces to the placer to swap router blocks for a certain percentage of the total number of swaps */
+
+/**
+ * @brief User supplied a number that represents the percentage of router 
+ * block swaps relative to the total number of swaps. A random number  
+ * between 0-100 is generated here and compared the user supplied value.
+ * If the random number is less than the user supplied value we indicate
+ * that a router block should be swapped. By doing this we now only swap
+ * router blocks for the percentage of time the user supplied.
+ * 
+ * @param user_supplied_noc_router_swap_percentage The expected number of 
+ * noc router block swaps as a percentage of the total number of swaps
+ * attempted by the placer.
+ * @return true This indicates that a router block should be swapped
+ * @return false This indicates that a router block should not be swapped
+ */
+bool check_for_router_swap(int user_supplied_noc_router_swap_percentage);
+
+/**
+ * @brief Generates a placement move by choosing two rouer cluster blocks to 
+ * swap. First, a random router cluster block is chosen and then another router
+ * cluster block is chosen that cqan be swapped with the initial block such that
+ * the distance travelled by either blocks does not exceed rlim. 
+ * 
+ * @param blocks_affected The two router cluster blocks that are proposed to be
+ * swapped
+ * @param rlim The maximum distance in the x and y direction that a router 
+ * cluster block can travel 
+ * @return e_create_move Result of proposing the move
+ */
+e_create_move propose_router_swap(t_pl_blocks_to_be_moved& blocks_affected, float rlim);
+
+/* Below are functions related to modifying and retreiving the NoC placement stats dastructure*/
+
+/**
+ * @brief Initializes all the stat values to 0 and allocates space for the
+ * number of possible move types offered by the placer. This is needed as
+ * the stats datastructure for the NoC keeps track of which moves caused
+ * a router block to move.
+ * 
+ * @param placer_opts Contains information about all the possible move
+ * types in the placer.
+ */
+void initialize_noc_placement_stats(const t_placer_opts& placer_opts);
```

Comment:
@saaramahmoudi : if we gathered move type stats per block type, this routine wouldn't be needed (well, you'd generalize it and use the general version for all blocks).

Reply:
@vaughnbetz Yes, I already added that in the placement stat, we don't need this.

Reply:
Sounds good. Thanks